### DISCUSSION TITLE
Understanding the relation of Taskchampion and Taskwarrior

### DISCUSSION THREAD

**Discussion by mrossinek** [taskchampion/taskchampion#395] [General]
Hi there! Avid Taskwarrior user and former contributor here :wave: 

I would like to understand the relation of Taskchampion (TC) and Taskwarrior (TW) a little bit better.
My understanding is, that TC was developed somewhat in parallel and for the major refactor 3.0 of TW, its data storage model (and sync) got integrated into TW.

The README of TC currently states:
> TaskChampion implements the task storage and synchronization behind Taskwarrior. It includes an implementation with Rust and C APIs, allowing any application to maintain and manipulate its own replica. It also includes a specification for tasks and how they are synchronized, inviting alternative implementations of replicas or task servers.

which matches my understanding described above.
That said, the landing page of the online documentation starts wheith:
> TaskChampion is a personal task-tracking tool. It works from the command line, with simple commands like `ta add "fix the kitchen sink"`.

which would lead to suggest that TC also includes a CLI in its own right, and I am trying to understand what the intention is here. Since it is deliberately stated that TC deviates slightly from the TW CLI, is TC for example one approach to write a newer CLI without the legacy of TW (as also briefly said in #349)?

EDIT: Apparently I did a _very_ poor job at checking other open issues and discussions before starting this thread.. I just found #392 which states that the docs that I quote above I severely out-of-date and that the CLI inside of TC actually no longer exists..
I am leaving this discussion here nonetheless, in the hope that it might still turn into a useful thread.

---

Providing some additional context: I am a very active TW user myself and have contributed in the past and am open to contributing to it and TC again in the future if time permits (that is).
That said, I am still using TW 2.x for the time being, because I do not want to give up on my self-hosted synchronization capabilities provided by the legacy `taskd` which also allows me to use [this TW 2.x compatible Android app](https://github.com/bgregos/foreground).

I am by no means a UI engineer and certainly not an Android App developer, but I am intrigued by the idea of having an interface to my synced tasks while on the go (even if it is more limited than the full CLI of course). In one of my own projects (a terminal-based bibliography manager, [coBib](https://gitlab.com/cobib/cobib)) I built a TUI using [textual](https://github.com/Textualize/textual/) a very powerful and fairly easy to use Python library. They are also experimenting with the ability to run Textual apps in a web browser (see [textual-web](https://github.com/Textualize/textual-web)) which is still in very early testing stages and has no self-hosting capabilities as of now, but I am cautiously optimistic that this could become a reality in the future.
Having seen #385 and #388 this got me really excited because it should make it fairly simple to code up a textual-based TUI client for the TC database which could (via `textual-web`) even be accessed from an arbitrary end device.

---

Okay, enough rambling. All the above is to say: I am excited to see more/fresh air breathed into the TW ecosystem and while I have not made the switch to TW 3.0 myself yet, I am excited for its potential with a more robust and potentially easier to work with data storage backend.

To summarize what I would like to here in this thread:
- what exactly is the intended future relation of TW and TC with their slightly differing CLIs (disclaimer: I did not look into the actual differences just yet)
- are others excited about the potential for different interfaces to a TC backend (I used Python as an example above, but of course, there are other options)

---

PS: shame on me, while ticking the `I have done a search for similar discussion` button, I discovered #272 which provides some more insight into some of the points I brought up here. That said, the last update there has been a while ago so maybe this new thread is still welcome and the title I used here might be easier to find by newcomers to the repo. If not, I am happy to continue discussion over in #272.

**Comment by djmitche** [comment]
Thanks! Yes, the docs are out of date - that's #392. There used to be a `ta` command-line tool, but it's gone now.

The lack of functional Android apps is part of what led me to work on this -- I currently use `task` in termux but reading the tiny text and trying to thumb out commands with no autocorrect is pretty annoying! Rust is now [a first-class citizen in the Android world](https://security.googleblog.com/2022/12/memory-safe-languages-in-android-13.html), so I'm hoping it's fairly easy to incorporate Taskchampion into a new Android app. Or at least, integrate it via JNI into a Java app. The only "supported" integration with Taskwarrior-2 was to execute the CLI, which wasn't designed for machine usage, and led to a lot of bugs.

The intended future relation between the two is that TW is the command-line tool and most powerful interface for tracking tasks, while TC is the data storage system that it uses and which can be used by other applications to interface with the same data.

It does seem others are interested in interfacing with TC, but there's some work to do to build and stabilize those interfaces.  Some of that work is scheduled for the [next milestone](https://github.com/orgs/GothenburgBitFactory/projects/4/views/1?filterQuery=status%3Aready).

We'd love to have your help! And we'd love to see you at the [next bi-weekly meeting](https://github.com/GothenburgBitFactory/taskwarrior/discussions/3477)!

**Reply by mrossinek** [reply]
Awesome thanks for the summary! Will see what I can do to chip in if time permits :+1: 
For now, I put this repo on my watch list to stay in the loop on what is happening.
With some time I will start peeking at the existing code and next issues. Maybe working on improving the docs would be a tough but good start as it would enforce getting familiar with the code. But I am not sure I have the time available for that atm.
